cpu/samd5x/periph/can.c

```diff
@@ -0,0 +1,777 @@
+    }
+    else if (dev->conf->can == CAN1) {
+        DEBUG_PUTS("CAN1 controller is used");
+        MCLK->AHBMASK.reg &= ~MCLK_AHBMASK_CAN1;
+    }
+    else {
+        DEBUG_PUTS("Unsupported CAN channel");
+        assert(0);
+    }
+
+    return 0;
+}
+
+static void _enter_init_mode(Can *can)
+{
+    can->CCCR.reg |= CAN_CCCR_INIT;
+    while (!(can->CCCR.reg & CAN_CCCR_INIT)) {}
+    DEBUG_PUTS("Device in init mode");
+}
+
+static void _exit_init_mode(Can *can)
+{
+    if (can->CCCR.reg & CAN_CCCR_INIT) {
+        can->CCCR.reg &= ~CAN_CCCR_INIT;
+    }
+
+    while (can->CCCR.reg & CAN_CCCR_INIT) {}
+    DEBUG_PUTS("Device out of init mode");
+}
+
+static int _set_mode(Can *can, can_mode_t can_mode)
+{
+    switch (can_mode) {
+        case MODE_INIT:
+            _enter_init_mode(can);
+            can->CCCR.reg |= CAN_CCCR_CCE;
+            break;
+        case MODE_TEST:
+            DEBUG_PUTS("test mode");
+            _enter_init_mode(can);
+            /* CCCR.TEST and CCCR.MON can be set only when CCCR.INIT and CCCR.CCE are set */
+            can->CCCR.reg |= CAN_CCCR_CCE;
+            can->CCCR.reg |= CAN_CCCR_TEST;
+            can->TEST.reg |= CAN_TEST_LBCK;
+#if IS_ACTIVE(CANDEV_SAMD5X_INTERNAL_LOOPBACK)
+            can->CCCR.reg |= CAN_CCCR_MON;
+#endif
+            _exit_init_mode(can);
+            break;
+        default:
+            DEBUG_PUTS("Unsupported mode");
+            return -1;
+    }
+
+    return 0;
+}
+
+static void _setup_clock(can_t *dev)
+{
+    if (dev->conf->can == CAN0) {
+        GCLK->PCHCTRL[CAN0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(dev->conf->gclk_src);
+    }
+    else if (dev->conf->can == CAN1) {
+        GCLK->PCHCTRL[CAN1_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(dev->conf->gclk_src);
+    }
+    else {
+        DEBUG_PUTS("CAN channel not supported");
+    }
+}
+
+static void _set_bit_timing(can_t *dev)
+{
+    assert(dev->candev.bittiming.sjw >= 1);
+    assert(dev->candev.bittiming.phase_seg2 >= 1);
+    assert(dev->candev.bittiming.phase_seg1 + dev->candev.bittiming.prop_seg >= 1);
+    assert(dev->candev.bittiming.brp >= 1);
+
+    DEBUG("bitrate=%" PRIu32 ", sample_point=%" PRIu32 ", brp=%" PRIu32 ", prop_seg=%" PRIu32
+          ", phase_seg1=%" PRIu32 ", phase_seg2=%" PRIu32 ", sjw=%" PRIu32 "\n", dev->candev.bittiming.bitrate,
+          dev->candev.bittiming.sample_point, dev->candev.bittiming.brp, dev->candev.bittiming.prop_seg,
+          dev->candev.bittiming.phase_seg1, dev->candev.bittiming.phase_seg2, dev->candev.bittiming.sjw);
+
+    /* Set bit timing */
+    dev->conf->can->NBTP.reg = (uint32_t)((CAN_NBTP_NTSEG2(dev->candev.bittiming.phase_seg2 - 1))
+                            | (CAN_NBTP_NTSEG1(dev->candev.bittiming.phase_seg1 + dev->candev.bittiming.prop_seg - 1))
+                            | (CAN_NBTP_NBRP(dev->candev.bittiming.brp - 1))
+                            | (CAN_NBTP_NSJW(dev->candev.bittiming.sjw - 1)));
+}
```

Comment:
there are still long lines may be this is a simple solution

```suggestion
static void _set_bit_timing(can_t *dev)
{
    struct can_bittiming *timing= &dev->candev.bittiming;
    
    assert(timing->sjw >= 1);
    assert(timing->phase_seg2 >= 1);
    assert(timing->phase_seg1 + timing.prop_seg >= 1);
    assert(timing->brp >= 1);

    DEBUG("bitrate=%" PRIu32 ", sample_point=%" PRIu32 ", brp=%" PRIu32 ", prop_seg=%" PRIu32
          ", phase_seg1=%" PRIu32 ", phase_seg2=%" PRIu32 ", sjw=%" PRIu32 "\n", timing->bitrate,
          timing->sample_point, timing->brp, timing->prop_seg,
          timing->phase_seg1, timing->phase_seg2, timing->sjw);

    /* Set bit timing */
    dev->conf->can->NBTP.reg = (uint32_t)((CAN_NBTP_NTSEG2(timing->phase_seg2 - 1))
                            | (CAN_NBTP_NTSEG1(timing->phase_seg1 + timing->prop_seg - 1))
                            | (CAN_NBTP_NBRP(timing->brp - 1))
                            | (CAN_NBTP_NSJW(timing->sjw - 1)));
}
```

Reply:
I fixed some of the warnings in the last commit, but left some too as splitting the lines will reduce a bit in the readability of the code IMO